File: scala/ch01/ex1.29.scala
def sum(term: (Int) => Double, m: Int, next: (Int)=>Int, n: Int): Double = 
  if (m > n) 0 else term(m) + sum(term, next(m), next, n)
    
def inc(n: Int) = n + 1
def even(n: Int) = n % 2 == 0

def simpsonsRule(f: (Double)=>Double, a: Double, b: Double, n: Int) = {
  val h = (b - a) / n
  def fy(k: Int) = f(a + (k * h))
  def term(i: Int) = fy(i) * (if (even(i)) 2 else 4)
  (h / 3) * (fy(0) + sum(term, 1, inc, (n - 1)) + fy(n))
}

val cube = (n: Double) => n * n * n

import org.scalatest._
import org.scalatest.matchers._

object simpsonsRuleSpec extends Spec with ShouldMatchers {
  describe ("simpsonsRule") {
    it ("should converge to the correct answer with more and more steps") {
      simpsonsRule(cube, 0.0, 1.0, 10)   should be (.25 plusOrMinus 0.01)
      simpsonsRule(cube, 0.0, 1.0, 100)  should be (.25 plusOrMinus 0.001)
      simpsonsRule(cube, 0.0, 1.0, 1000) should be (.25 plusOrMinus 0.0001)
      // Stack overflow (but see 1.30)
      // simpsonsRule(cube, 0.0, 1.0, 10000) should be (.25 plusOrMinus 0.00001)
    }
  }
}

simpsonsRuleSpec execute

File: clojure/ch01/ex1.29.clj
(ns sicp.ch01 (:use clojure.contrib.test-is))

(defn sum [term a next b]
  (if (> a b)
    0
    (+ (term a) (sum term (next a) next b))))
    
(defn inc [n] (+ n 1))
(defn even? [n] (= 0 (rem n 2)))

(defn simpsons-rule [f a b n]
  (def h (/ (- b a) n))
  (defn fy [k] (f (+ a (* k h))))
  (defn term [i]
    (* (fy i) (if (even? i) 2 4)))
  (* (/ h 3) (+ (fy 0) (sum term 1 inc (- n 1)) (fy n))))
  
(defn cube [n] (* n n n))

(defn abs [x] (if (> x 0.0) x (- x)))

(deftest test-simpsons-rule
  (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 10)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 100)))))
  (is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 1000)))))
  ; Appears to stack overflow; see next problem that uses tail recursion
  ;(is (> 0.001 (abs (- 0.25 (simpsons-rule cube 0.0 1.0 10000)))))
  )
